File: python/cmpad/pytorch/llsq_obj.py
# SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later
# SPDX-FileCopyrightText: Bradley M. Bell <bradbell@seanet.com>
# SPDX-FileContributor: 2023-24 Bradley M. Bell
# ---------------------------------------------------------------------------
r'''
{xrst_begin pytorch_llsq_obj}

Pytorch Linear Least Squares Objective
######################################

Syntax
******
| |tab| *llsq* = ``cmpad.pytorch.llsq_obj()``
| |tab| *llsq* . ``setup`` ( *option* )
| |tab| *obj* = *llsq* ( *x* )

Prototype
*********
{xrst_literal
   # BEGIN PROTOTYPE
   # END PROTOTYPE
}

Special
*******
This is a python implementation of the :ref:`llsq_obj-name` algorithm
that is specialized for the pytorch package.

n_arg
*****
see :ref:`llsq_obj@option@n_arg` .

n_other
*******
see :ref:`llsq_obj@option@n_other` .

Source Code
***********
The code below is the implementation of this function:
{xrst_literal
   # BEGIN PYTHON
   # END PYTHON
}


{xrst_end pytorch_llsq_obj}
'''
# BEGIN PYTHON
import numpy
import torch
#
# BEGIN PROTOTYPE
class llsq_obj :
   def domain(self) :
      return self.n_arg
   def range (self) :
      return 1
   #
   def setup(self, option) :
      assert type(option) == dict
      assert type(option['n_arg']) == int
      assert type(option['n_other']) == int
      assert option['n_arg'] > 0
      assert option['n_other'] > 0
      # END PROTOTYPE
      #
      # n_arg, n_other
      n_arg   = option['n_arg']
      n_other = option['n_other']
      #
      # t
      if n_other == 1 :
         t = numpy.array( [0.0 ] )
      else :
         t = numpy.linspace(-1.0, 1.0, n_other)
      #
      # q
      q = -1.0 * (t < 0) + 1.0 * (t > 0)
      #
      # self
      self.n_arg = n_arg
      self.t     = torch.tensor(t, dtype=float)
      self.q     = torch.tensor(q, dtype=float)
   #
   def __call__(self, ax) :
      assert len(ax) == self.n_arg
      #
      # model
      model = 0.0
      ti  = torch.ones( len(self.t) , dtype=float)
      for i in range(self.n_arg) :
         model = model + ax[i] * ti
         ti    = ti * self.t
      #
      # squared_residual
      squared_residual = torch.square(model - self.q)
      #
      # objective
      objective = 0.5 * squared_residual.sum()
      #
      return objective.view(1)
# END PYTHON

File: cpp/xam/cppad/llsq_obj.cpp
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later
// SPDX-FileCopyrightText: Bradley M. Bell <bradbell@seanet.com>
// SPDX-FileContributor: 2023-24 Bradley M. Bell
// ----------------------------------------------------------------------------

/*
{xrst_begin xam_cppad_llsq_obj.cpp}

Example and Test of CppAD Special llsq_obj
##########################################

{xrst_literal
   // BEGIN C++
   // END C++
}

{xrst_end xam_cppad_llsq_obj.cpp}
*/
// BEGIN C++
# include <cmpad/cppad/llsq_obj.hpp>
# include <cmpad/near_equal.hpp>
# include <cppad/example/valvector/class.hpp>

bool xam_cppad_llsq_obj(void)
{  // ok
   bool ok = true;
   //
   // scalar_type
   typedef CppAD::AD<valvector> scalar_type;
   //
   // llsq
   typedef typename cmpad::vector<scalar_type> Vector;
   cmpad::cppad::llsq_obj<Vector>              llsq;
   //
   // n_arg, n_other
   size_t n_arg   = 3;
   size_t n_other = 4;
   //
   // llsq.setup
   cmpad::option_t option;
   option.n_arg   = n_arg;
   option.n_other = n_other;
   llsq.setup(option);
   //
   // ax
   cmpad::vector<double> x(n_arg);
   Vector ax(n_arg);
   for(size_t i = 0; i < n_arg; ++i)
   {  x[i]  = double(i);
      ax[i] = x[i];
   }
   //
   // ay
   const Vector& ay = llsq(ax);
   ok &= ay.size()    == 1;
   //
   // sumsq
   double sumsq = 0.0;
   //
   for(size_t j = 0; j < n_other; ++j)
   {  //
      // t_j
      double t_j = -1.0 + 2.0 * double(j) / double(n_other - 1);
      //
      // q_j
      double q_j = std::copysign( 1.0, t_j );
      if( t_j == 0 )
         q_j = 0.0;
      //
      // r_j
      double m_j = x[0] + x[1] * t_j + x[2] * t_j * t_j;
      double r_j = m_j - q_j;
      //
      // sumsq
      sumsq += r_j * r_j;
   }
   //
   // check
   double check = 0.5 * sumsq;
   //
   // rel_error
   double rel_error = std::numeric_limits<double>::epsilon() * 100.0;
   //
   // ok
   valvector y0 = Value( ay[0] );
   ok &= cmpad::near_equal( y0[0], check, rel_error );
   //
   return ok;
}

// END C++
